src/basic/build-path.c

```diff
@@ -45,7 +46,19 @@ static int get_runpath_from_dynamic(const ElfW(Dyn) *d, const char **ret) {
         if (!strtab)
                 return -ENOTRECOVERABLE;
 
-        /* According to dl.so runpath wins of both runpath and rpath are defined. */
+        /* On MIPS and RISC-V DT_STRTAB records an offset, not a valid address, so it has to be adjusted.
+         * Check whether strtab is below the PT_DYNAMIC, and if so assume it's an offset. */
+#if defined(__mips__) || defined(__riscv)
+        struct link_map *map;
+        Dl_info info;
+
+        if (dladdr1(dynamic, &info, (void **)&map, RTLD_DL_LINKMAP) <= 0)
+                return -ENOTRECOVERABLE;
+
+        strtab += map->l_addr;
```

Comment:
hmm, isn't what you are looking for here just the `bias` value we calculated in get_runpath() (i.e. one stack frame further up) already? dladdr1() is kinda slow, passing in `bias` here seems a lot nicer (and shorter)

Reply:
yeah that seems to work equally well